Feature: BB-146 support different internal topics for notification targets

welansari opened the pull request:
Issue: [BB-146](https://scality.atlassian.net/browse/BB-146)

**Changes**:

NotificationQueuePopulator:
- Now pushes notification messages onto specific internal Kafka topics of destinations for which the object validates their notification config.

NotificationQueueProcessor:
- Now consumes messages form internal Kafka topic specific to their destination
- Made processKafkaEntry use callbacks to fix an async bug where we get the following error `done is not a function`

Config
- Added fields in the notification config to specify the internal topic to use for each destination

ReplicationProcessor:
- Now supports pushing replication failed messages to specific internal topic of destinations

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-146 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.2`


Please check the `Fix Version/s` of BB-146, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-146 contains:


* `8.4.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.3`


Please check the `Fix Version/s` of BB-146, or the target
branch of this pull request.

francoisferrand commented on extensions/notification/NotificationConfigValidator.js:
```diff
@@ -1,7 +1,12 @@
+
 const joiSchema = joi.object({
-    topic: joi.string(),
+    topics: joi.array().items(internalTopic).required(),
```
~should still support the "old" format with just a single stream?~

or actually, the topic should be part of each "target"; and this global `topic` config may be used as the default or prefix
→ i.e. allow specifying the name of the "internal" topic to use inside each target, else use some default

basically it seems quite redundant and risky to require "matching" different parts of the config (and not very intuitive)

francoisferrand commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -73,21 +73,38 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
-            }
-            return undefined;
+            // validate and push kafka message foreach destination topic
+            this.notificationConfig.topics.forEach(destination => {
```
seems this is not "topics" but "destinations" ?

francoisferrand commented on extensions/notification/queueProcessor/QueueProcessor.js:
```diff
@@ -45,6 +47,12 @@ class QueueProcessor extends EventEmitter {
         this.bnConfigManager = null;
         this._consumer = null;
         this._destination = null;
+        /**
+         * Once the notification manager is initialized
+         * this will hold the callback version of the getConfig
+         * function of the notification config manager
+         */
```
should not be a doc comment?

```suggestion
         // Once the notification manager is initialized
         // this will hold the callback version of the getConfig
         // function of the notification config manager
```

Commit:
feature: BB-146 update notification config to contain destination configs

Commit:
feature: BB-146 allow pushing to different topics in notification queue populator

NotificationQueuePopulator now only pushes messages to the destinations' internal
topics that are now specified in the destination config

If no internal topic is specified for a destination it falls back on the default
internal topic

Commit:
feature: BB-146 consume from specific topic in the notification QueueProcessor

The NotificationQueueProcessor now consumes from its destination's internal topic
if specified in the config, otherwise it consumes form the default internal topic

Commit:
feature: BB-146 remove async from processKafkaEntry

Make processKafkaEntry use callbacks instead of async/await
to avoid getting async errors "done is not a function"

Commit:
feature: BB-146 fix NotificationQueuePopulator unit tests

Commit:
feature: BB-146 support pushing replication failed notification to multiple topics

Updated the replication workflow to allow pushing replication failed notifications
to specific internal topics used by the notification destinations

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-146 contains:


* `8.4.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.4`


Please check the `Fix Version/s` of BB-146, or the target
branch of this pull request.

jonathan-gramain commented on extensions/notification/queueProcessor/QueueProcessor.js:
```diff
@@ -28,23 +30,28 @@ class QueueProcessor extends EventEmitter {
         this.destinationId = destinationId;
+        this.destinationConfig
+            = notifConfig.destinations.find(dest => dest.resource === destinationId);
+        assert(this.destinationConfig, 'Invalid destination argument. Destination ' +
```
It's minor as this is only to debug a config issue, but we could help the admin by logging the `destinationId` with a formatted string

jonathan-gramain commented on extensions/notification/queueProcessor/QueueProcessor.js:
```diff
@@ -158,59 +170,69 @@ class QueueProcessor extends EventEmitter {
+                        });
+                        return this._destination.send([msg], done);
+                    }
+                    return done();
```
This line looks unnecessary, as the callback will be called at 234

welansari commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-146 contains:


* `8.4.2`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.4`


Please check the `Fix Version/s` of BB-146, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`



Please check the status of the associated issue BB-146.

Goodbye kerkesni.
